pkg/reconciler/workload/reconciler.go

```diff
@@ -0,0 +1,188 @@
+	applicator  resource.Applicator
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Kind is a kind of OAM workload.
+type Kind schema.GroupVersionKind
+
+// A Workload is a type of OAM workload.
+type Workload interface {
+	resource.Conditioned
+	metav1.Object
+	runtime.Object
+}
+
+// An Object is a Kubernetes object.
+type Object interface {
+	metav1.Object
+	runtime.Object
+}
+
+// NewReconciler returns a Reconciler that reconciles an OAM workload type by
+// packaging it into a KubernetesApplication.
+func NewReconciler(m ctrl.Manager, workload Kind, o ...ReconcilerOption) *Reconciler {
+	nw := func() Workload {
+		return resource.MustCreateObject(schema.GroupVersionKind(workload), m.GetScheme()).(Workload)
+	}
+
+	r := &Reconciler{
+		client:      m.GetClient(),
+		newWorkload: nw,
+		workload:    TranslateFn(NoopTranslate),
+		applicator:  resource.ApplyFn(resource.Apply),
+		log:         logging.NewNopLogger(),
+		record:      event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile an OAM workload type by packaging it into a KubernetesApplication.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	workload := r.newWorkload()
+	if err := r.client.Get(ctx, req.NamespacedName, workload); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetWorkload)
+	}
+
+	log = log.WithValues("uid", workload.GetUID(), "version", workload.GetResourceVersion())
+
+	objs, err := r.workload.Translate(ctx, workload)
+	if err != nil {
+		log.Debug("Cannot translate workload", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(workload, event.Warning(reasonCannotTranslateWorkload, err))
+		workload.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errTranslateWorkload)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, workload), errUpdateWorkloadStatus)
+	}
+
+	for _, o := range objs {
+		if err := r.applicator.Apply(ctx, r.client, o, resource.ControllersMustMatch()); err != nil {
+			log.Debug("Cannot apply workload translate", "error", err, "requeue-after", time.Now().Add(shortWait))
+			r.record.Event(workload, event.Warning(reasonCannotApplyWorkloadTranslation, err))
+			workload.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errApplyWorkloadTranslation)))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, workload), errUpdateWorkloadStatus)
+		}
+
+		// A workload's translation must be controlled by the workload.
+		meta.AddOwnerReference(o, *metav1.NewControllerRef(workload, workload.GetObjectKind().GroupVersionKind()))
+
+		// All top-level objects must be in the same namespace as the workload.
+		o.SetNamespace(workload.GetNamespace())
```

Comment:
This has the effect of only setting an owner reference and namespace on the "outer-most layer" of the translation. This is important for the remote vs local case:

1. Remote: only one object comes out of the translation (i.e. a `KubernetesApplication`). We do not want the embedded resources to be controlled by the workload.
2. Local: multiple objects (i.e. `Deployment` and `Service`) come out of the translation. We want both to be owned by the workload.

We also want to ensure that created objects always reside in the namespace of the workload, so we set it in the reconciler rather than the translation.